File: ISA-CMOP_Python/features/dist_corr.py
from features.feature_helpers import remove_imag_rows, corr_coef
from scipy.spatial.distance import cdist
import numpy as np


def dist_corr(pop, NonDominated):
    """
    Distance correlation.

    Distance for each solution to nearest global solution in decision space. Correlation of distance and constraints norm.

    Fairly certain that NonDominated is an instance of the Population class containing the non-dominated solutions
    """

    obj = pop.extract_obj()
    var = pop.extract_var()
    cons = pop.extract_cons()  # constraint matrix.
    cv = pop.extract_cv()

    # Remove imaginary rows. Deep copies are created here.
    obj = remove_imag_rows(obj)
    var = remove_imag_rows(var)
    cons = remove_imag_rows(cons)

    # For each ND decision variable, find the smallest distance to the nearest population decision variable.
    dist_matrix = cdist(NonDominated.extract_var(), var, "euclidean")
    min_dist = np.min(dist_matrix, axis=0)

    # Then compute correlation coefficient between CV and . Assumed that all values in the correlation matrix are the same, meaning we only need one scalar. See f_corr for a similar implementation.
    dist_c_corr = corr_coef(cv, min_dist)
    return [dist_c_corr]

File: ISA-CMOP_Python/testing/dist_corr.m
function [dist_c_corr] = dist_corr(pop, NonDominated)
%distance correlation
%distance for each solution to nearest global solution in the decsion space
%correlation of distance and constraints norm. 

    objvar = pop.objs; decvar = pop.decs; consvar = pop.cons;
	rmimg = find(sum(imag(objvar)~= 0,2));
	objvar(rmimg,:) = [];
	%decvar(rmimg,:) = [];
    consvar(rmimg,:) = [];
    
    dist = pdist2(NonDominated.decs,decvar,'euclidean','Smallest',1);
    [c1,pvalue] = corrcoef(pop.normcvs,dist.');
    dist_c_corr = c1(1,2);
    if pvalue(1,2) > 0.05 %there is no corr 
        dist_c_corr = 0;
    end
    
end

